Content.Shared/Labels/EntitySystems/SharedHandLabelerSystem.cs

```diff
@@ -59,64 +59,82 @@ protected virtual void UpdateUI(Entity<HandLabelerComponent> ent)
     {
     }
 
-    private void AddLabelTo(EntityUid uid, HandLabelerComponent? handLabeler, EntityUid target, out string? result)
+    private void AddLabelTo(Entity<HandLabelerComponent> ent, EntityUid user, EntityUid target)
     {
-        if (!Resolve(uid, ref handLabeler))
+        if (ent.Comp.AssignedLabel == string.Empty)
         {
-            result = null;
+            RemoveLabelFrom(ent, user, target);
             return;
         }
 
-        if (handLabeler.AssignedLabel == string.Empty)
-        {
-            if (_netManager.IsServer)
-                _labelSystem.Label(target, null);
-            result = Loc.GetString("hand-labeler-successfully-removed");
-            return;
-        }
         if (_netManager.IsServer)
-            _labelSystem.Label(target, handLabeler.AssignedLabel);
-        result = Loc.GetString("hand-labeler-successfully-applied");
+            _labelSystem.Label(target, ent.Comp.AssignedLabel);
+
+        _popupSystem.PopupClient(Loc.GetString("hand-labeler-successfully-applied"), user, user);
+
+        // Log labeling
+        _adminLogger.Add(LogType.Action, LogImpact.Low,
+            $"{ToPrettyString(user):user} labeled {ToPrettyString(target):target} with {ToPrettyString(ent):labeler}");
+    }
+
+    private void RemoveLabelFrom(EntityUid uid, EntityUid user, EntityUid target)
+    {
+        if (_netManager.IsServer)
+            _labelSystem.Label(target, null);
+
+        _popupSystem.PopupClient(Loc.GetString("hand-labeler-successfully-removed"), user, user);
+
+        // Log labeling
+        _adminLogger.Add(LogType.Action, LogImpact.Low,
+            $"{ToPrettyString(user):user} removed label from {ToPrettyString(target):target} with {ToPrettyString(uid):labeler}");
     }
 
     private void OnUtilityVerb(EntityUid uid, HandLabelerComponent handLabeler, GetVerbsEvent<UtilityVerb> args)
     {
         if (args.Target is not { Valid: true } target || _whitelistSystem.IsWhitelistFail(handLabeler.Whitelist, target) || !args.CanAccess)
             return;
 
-        var labelerText = handLabeler.AssignedLabel == string.Empty ? Loc.GetString("hand-labeler-remove-label-text") : Loc.GetString("hand-labeler-add-label-text");
+        bool labelerBlank = (handLabeler.AssignedLabel == string.Empty);
```

Comment:
Can't mark line 92 but make this method `private void OnUtilityVerb(Entity<HandLabelerComponent> ent, ref GetVerbsEvent<UtilityVerb> args)`